common/rfb/CSecurityRSAAES.cxx

```diff
@@ -0,0 +1,444 @@
+/* 
+ * Copyright (C) 2022 Dinglan Peng
+ *    
+ * This is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ * 
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public License
+ * along with this software; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
+ * USA.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#ifndef HAVE_GNUTLS
+#error "This header should not be compiled without HAVE_GNUTLS defined"
+#endif
+
+#include <stdlib.h>
+#ifndef WIN32
+#include <unistd.h>
+#endif
+
+#include <gnutls/gnutls.h>
+#include <gnutls/crypto.h>
+#include <gnutls/x509.h>
+#include <nettle/bignum.h>
+#include <nettle/sha1.h>
+#include <nettle/sha2.h>
+#include <rfb/CSecurityRSAAES.h>
+#include <rfb/CSecurityTLS.h>
+#include <rfb/CConnection.h>
+#include <rfb/LogWriter.h>
+#include <rfb/Exception.h>
+#include <rfb/UserMsgBox.h>
+#include <rdr/AESInStream.h>
+#include <rdr/AESOutStream.h>
+#include <os/os.h>
+
+enum {
+  ReadPublicKey,
+  ReadRandom,
+  ReadHash,
+  ReadSubtype,
+};
+
+const int MinKeyLength = 1024;
+const int MaxKeyLength = 8192;
+
+using namespace rfb;
+
+IntParameter CSecurityRSAAES::RSAKeyLength
+("RSAAESKeyLength", "bits of the key generated for RSA-AES security types",
+  2048, 1024, 8192, ConfViewer);
+
+
+CSecurityRSAAES::CSecurityRSAAES(CConnection* cc, rdr::U32 _secType)
+  : CSecurity(cc), state(ReadPublicKey), secType(_secType),
+    clientKey{0}, clientPublicKey{0}, serverKey{0},
+    serverKeyN(NULL), serverKeyE(NULL),
+    clientKeyLength(RSAKeyLength), clientKeyN(NULL), clientKeyE(NULL),
+    rais(NULL), raos(NULL), rawis(NULL), rawos(NULL)
+{
+  rawis = cc->getInStream();
+  rawos = cc->getOutStream();
+  if (gnutls_global_init() != GNUTLS_E_SUCCESS)
+    throw AuthFailureException("gnutls_global_init failed");
```

Comment:
Hmm... I think it is excessive to require GnuTLS just to have a random function. Have you looked at the `RandomStream` class?

Reply:
I think that can work as a start. But I think PEM files are just DER files with Base64 encoding, so it would hopefully not be too difficult to write a small conversion function that doesn't need GnuTLS.